### DISCUSSION TITLE
Question: Constraint engine?

### DISCUSSION THREAD

**Discussion by bhouston** [elalish/manifold#118] [Q&A]
I wonder if it is possible to integrate a 2D constraint engine with this library?  That would be a very key step in moving towards proper CAD.

I wonder what FreeCAD uses for its 2D constraint system?  https://wiki.freecadweb.org/Sketcher_Workbench

**Comment by elalish** [comment]
Yes, although if what you want is traditional CAD, then you probably want NURBS and other feature-based flows. The idea of doing this in triangle meshes is that you can mix up the Booleans with sculpting and SDFs and generative design. A constraints engine would probably serve as an earlier stage: creating meshes and transforms for the inputs.

**Reply by rsaccon** [reply]
I don't see the benefit of integrating it. I would use an existing constraint engine like [Solvespace](https://solvespace.github.io/solvespace-web/library.html) to create the Polygons.

**Comment by jirihon** [comment]
I think it's very possible to integrate 2D constraint engine with manifold. I personally tried the solver used in FreeCAD which is called [planegcs](https://github.com/FreeCAD/FreeCAD/tree/master/src/Mod/Sketcher/App/planegcs). The solver can be easily compiled independently. I was able to compile it into WebAssembly module and use it on the web together with manifold. Eventually, I'll publish the js bindings when I figure out the right shape for the interface.

**Reply by elalish** [reply]
Excellent! Keep us posted, I'd love to see how it works.

**Reply by NodeGuy** [reply]
Hi @jirihon have you published anything yet?  I'm very interested in this.

**Reply by jirihon** [reply]
Hi @NodeGuy and @bhouston, if you are ok with no documentation and high chance of breaking changes, we now have a public repo of our npm package here: https://github.com/Salusoft89/planegcs It is not yet in npm, so you need to build it yourself using `npm install && npm run build` (requires docker). The license is LGPL, same as for the original C++ code, which makes it the only fully-featured 2D geometric constraint solver usable for closed-source applications.

It should cover all the functionality provided by the planegcs solver from FreeCAD as the JS bindings are generated automatically from the original C++ code. However, we are still learning how to use the solver properly. There are many hidden pitfalls. Certain combinations of constraints might not converge well and FreeCAD has some higher-level logic around the solver to choose the right constraints in the right situations. E.g. when you have two adjacent curves which are tangent, it is better to model it with zero angle constraint between the two curves in the common point than with a tangent constraint. So, right now, we are not experienced enough to give support how to use it.

@NodeGuy thanks for the link to the new [slvs](https://www.npmjs.com/package/slvs/v/3.1.0-dev.22) package. It looks nice, just the GPL license limits its applicability.

**Reply by elalish** [reply]
Thanks for sharing!
